subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1727 @@
+			}
+			*addr = ate_end_addr;
+		}
+		ate_end_addr -= ate_size;
+	}
+
+	return 0;
+}
+
+/* compute previous addr of ATE */
+static int zms_compute_prev_addr(struct zms_fs *fs, uint64_t *addr)
+{
+	int sec_closed;
+	struct zms_ate empty_ate, close_ate;
+	size_t ate_size;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	*addr += ate_size;
+	if ((SECTOR_OFFSET(*addr)) != (fs->sector_size - 2 * ate_size)) {
+		return 0;
+	}
+
+	/* last ate in sector, do jump to previous sector */
+	if (SECTOR_NUM(*addr) == 0U) {
+		*addr += ((uint64_t)(fs->sector_count - 1) << ADDR_SECT_SHIFT);
+	} else {
+		*addr -= (1ULL << ADDR_SECT_SHIFT);
+	}
+
+	/* verify if the sector is closed */
+	sec_closed = zms_validate_closed_sector(fs, *addr, &empty_ate, &close_ate);
+	if (sec_closed < 0) {
+		return sec_closed;
+	}
+
+	/* Non Closed Sector */
+	if (!sec_closed) {
+		/* at the end of filesystem */
+		*addr = fs->ate_wra;
+		return 0;
+	}
+
+	/* Update the address here because the header ATEs are valid.*/
+	(*addr) &= ADDR_SECT_MASK;
+	(*addr) += close_ate.offset;
+
+	return 0;
+}
+
+/* walking through allocation entry list, from newest to oldest entries
+ * read ate from addr, modify addr to the previous ate
+ */
+static int zms_prev_ate(struct zms_fs *fs, uint64_t *addr, struct zms_ate *ate)
+{
+	int rc;
+	size_t ate_size;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	rc = zms_flash_ate_rd(fs, *addr, ate);
+	if (rc) {
+		return rc;
+	}
+
+	return zms_compute_prev_addr(fs, addr);
+}
+
+static void zms_sector_advance(struct zms_fs *fs, uint64_t *addr)
+{
+	*addr += (1ULL << ADDR_SECT_SHIFT);
+	if ((*addr >> ADDR_SECT_SHIFT) == fs->sector_count) {
+		*addr -= ((uint64_t)fs->sector_count << ADDR_SECT_SHIFT);
+	}
+}
+
+/* allocation entry close (this closes the current sector) by writing offset
+ * of last ate to the sector end.
+ */
+static int zms_sector_close(struct zms_fs *fs)
+{
+	int rc;
+	struct zms_ate close_ate, garbage_ate;
+	size_t ate_size;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	close_ate.id = ZMS_HEAD_ID;
+	close_ate.len = 0U;
+	close_ate.offset = (uint32_t)SECTOR_OFFSET(fs->ate_wra + ate_size);
+	close_ate.metadata = 0xffffffff;
+	close_ate.cycle_cnt = fs->sector_cycle;
+
+	/* When we close the sector, we must write all non used ATE with
+	 * a non valid (Junk) ATE.
+	 * This is needed to avoid some corner cases where some ATEs are
+	 * not overwritten and become valid when the cycle counter wrap again
+	 * to the same cycle counter of the old ATE.
+	 * Example :
+	 * - An ATE.cycl_cnt == 0 is written as last ATE of the sector
+	   - This ATE was never overwritten in the next 255 cycles because of
+	     large data size
+	   - Next 256th cycle the leading cycle_cnt is 0, this ATE becomes
+	     valid even if it is not the case.
+	 */
+	memset(&garbage_ate, fs->flash_parameters->erase_value, sizeof(garbage_ate));
+	while (SECTOR_OFFSET(fs->ate_wra) && (fs->ate_wra >= (fs->data_wra + ate_size))) {
```

Comment:
Actually, shouldn't the condition be:
```suggestion
	while (SECTOR_OFFSET(fs->ate_wra) && (fs->ate_wra >= fs->data_wra)) {
```
`ate_wra` points to the to-be-added ATE so even if it's equal to `data_wra` there's still unused ATE, right?

Reply:
Yes indeed, will fix this